### ISSUE TITLE
Proposal: `mix igniter.gen.ci [github]` to generate GitHub actions config

### ISSUE THREAD

**Issue by zachallaun** [ash-project/igniter#119]
Whenever I start a new project, one of the first things I do is copy the GitHub actions config from another project over to the new one. An Igniter generator that does this would be quite nice.

Some quick thoughts:

- Have the config type (GitHub, GitLab, Circle, etc.) be a positional arg that defaults to `github`. Initial implementation can just support `github`, but can be written with others in mind.
- Set up a `test` job
- Set up a `lint` job that includes `mix format --check-formatted`, `mix deps.unlock --check-unused`, `mix compile --warnings-as-errors`
  - If `:dialyxir` is installed run Dialyzer in the `lint` job and cache PLTs

**Comment by zachdaniel** [comment]
Definitely a very interested concept! I'd love to standardize on this sort of "more advanced" setup for a new project. A lot of thought will need to go into it, but I think it would be worth considering standardizing even on a set of shared CI tasks that folks could use, or even a shared CI source? For instance, all of our projects point at `ash-ci.yml` in the ash repo.

This is *probably* taking things a bit too far given that people will often end up needing to customize their CI, but its been amazing for Ash projects to be able to just add new CI steps (like when we realized we needed `mix hex.audit` etc.) in just one place, and automatically get picked up in CI.

But....yeah I've already talked myself out of it 😆 Just having a starting point that does the major things would be the way.

Some things we should consider:

- [ ] `mix check` - it is a very useful tool for new projects to have, but I don't like using it in CI personally because you don't get separate job statuses.
- [ ] Should we provide a good starting point for:
  - [ ] `sobelow`
  - [ ] `mix_audit` (check for CVEs for hex packages)
  - [ ] `credo`
  - [ ] more tools I'm not thinking of

If we were to standardize on `ex_check`, then when packages install themselves, they can add recommended checks into the `.check.exs`. If there was then some way for that `.check.exs` to be authoritative in CI that *is not* just run `mix check` (for the reasons I described above), that would be pretty epic. We'd have a standardized way for libraries to patch checks into both a users local commands `mix check` and CI.

**Comment by zachdaniel** [comment]
The *default* `mix igniter.new` should likely remain as is, and any additional behavior here would be added with another generator, or something like `--ci <platform1>,<platform2>` when making a new project potentially?

**Comment by zachallaun** [comment]
Regarding the more advanced setup, I think that, for now, having the starting point is best.

Regarding the inclusion of other tools, I don't think `mix igniter.gen.ci` should add any deps by default, but if you already have them, it would be *amazing* to include their usage by default in the `lint` job! Since `mix check` is a sort of catch-all, maybe there's a special case where, if `ex_check` is installed, `mix check` is used instead of a bunch of separate commands?

Perhaps additional deps could be added and those checks included with a flag like `--include sobelow,credo`?

> The _default_ `mix igniter.new` should likely remain as is, and any additional behavior here would be added with another generator, or something like `--ci <platform1>,<platform2>` when making a new project potentially?

Agreed here!

**Comment by zachdaniel** [comment]
Thinking about it, I think maybe we should let this be "someone else's problem". Like let's convince the ex_check folks to make an igniter installer that generates high quality CI files in addition to a `.check.exs`. Or some other package/tool that makes life easier. Then we can just `mix igniter.new foo --install ex_check` etc.

I think later on down the road we're going to want to solve the "bundling" problem to let people build starter commands that compose many installers.

Regardless, I think we're providing all the tools for something like what you've described to be built, and maybe its best if we just let someone else use those tools to build it?

**Comment by zachallaun** [comment]
> Regardless, I think we're providing all the tools for something like what you've described to be built, and maybe its best if we just let someone else use those tools to build it?

If we ignore most of the complexity above (handling various dependencies and generating nearly-optimal CI stuff for them), I think there's a lot of benefit to having a fairly bare-bones generator in Igniter itself. Speaking for myself, I probably wouldn't use `ex_check`, but would use the heck out of `mix igniter_new my_new_project --ci github`.

**Comment by zachdaniel** [comment]
Right, I'm suggesting maybe we find a way to let "whatever that is" be managed outside of igniter. Like igniter shouldn't necessarily be the authority on what a good GitHub CI file looks like. So maybe there could be packages that you install that are on hex but are removed after you install them. So they exist only to generate code one time. Like `mix igniter.install github_ci` or something.

Then that project can grow separately from our own.

**Comment by zachallaun** [comment]
> So maybe there could be packages that you install that are on hex but are removed after you install them. So they exist only to generate code one time.

That's an interesting thought! I definitely think that's a pattern that could work.

I likely won't pick up the torch to do that for this CI stuff (maintain a separate package), but I think it's a fine direction for this issue should someone take it up.

**Comment by zachallaun** [comment]
Will close this for now and it can be perhaps reopened in the future if someone wants to take this on and is interested in further discussion on this direction!
